server/src/Persistence.hs

```diff
@@ -69,9 +76,94 @@ apply op val = do
+
+-- * SQLite loading and syncing
+
+-- | There is currently just one row and it contains only one column of type SBS.ByteString.
+-- This single field holds the whole JSON value for now.
+data JsonRow = JsonRow {jsonByteString :: SBS.ByteString} deriving (Show)
+
+instance FromRow JsonRow where
+  fromRow = JsonRow <$> field
+
+loadSqliteFile :: PersistenceConfig -> IO (Either String PersistentValue)
+loadSqliteFile config = runExceptT $ do
+  liftIO $ forM_ (pcMetrics config) $ \m -> do
+    size <- getFileSize (pcDataFile config)
+    Metrics.setDataSize size m
+
+  value <- readSqliteData (pcDataFile config) (pcLogger config)
+  valueVar <- lift $ newTVarIO value
+  dirtyVar <- lift $ newTVarIO False
+  journal <- for (pcJournalFile config) openJournal
+  let val = PersistentValue
+        { pvConfig  = config
+        , pvValue   = valueVar
+        , pvIsDirty = dirtyVar
+        , pvJournal = journal
+        }
+  recoverJournal val
+  return val
+
+-- | Read and decode the Sqlite data file from disk
+readSqliteData :: FilePath -> Logger -> ExceptT String IO Store.Value
+readSqliteData filePath _logger = ExceptT $ do
+  -- read the data from SQLite
+  conn <- liftIO $ open filePath
+  -- TODO: We might want to have a primary key here after all. Then we could have multiple
+  -- JSON values next to each other.
+  liftIO $ execute_ conn "CREATE TABLE IF NOT EXISTS test (value BLOB)"
+  jsonRows <- liftIO $ (query_ conn "SELECT * from test" :: IO [JsonRow])
+
+  case jsonRows of
+    -- if there is no data yet, we simply return the empty object. We do the same thing for the
+    -- file backend.
+    [] -> do
+      liftIO $ execute conn "INSERT INTO test (value) VALUES (?)" (Only $ Aeson.encode Aeson.emptyObject)
+      pure $ Right Aeson.emptyObject
+    _  -> case Aeson.eitherDecodeStrict (jsonByteString $ head $ jsonRows) of
+            Left msg  -> pure $ Left $ "Failed to decode the initial data: " ++ show msg
+            Right value -> pure $ Right (value :: Store.Value)
+
+-- | Write the data to the SQLite file if it has changed.
+syncSqliteFile :: PersistentValue -> IO ()
+syncSqliteFile val = do
+  (dirty, value) <- atomically $ (,) <$> readTVar (pvIsDirty val)
+                                     <*> readTVar (pvValue val)
+                                     <*  writeTVar (pvIsDirty val) False
+  -- simple optimization: only write when something changed
+  when dirty $ do
+    let filePath = pcDataFile $ pvConfig val
+
+    conn <- open filePath
+    -- we can always UPDATE here, since we know that there will be at least one row, since
+    -- we issue an INSERT when we load in an empty database
+    liftIO $ executeNamed conn "UPDATE test SET value = :value" [":value" := Aeson.encode value]
+
+    -- the journal is idempotent, so there is no harm if icepeak crashes between
+    -- the previous and the next action
+    for_ (pvJournal val) $ \journalHandle -> do
+      hSeek journalHandle AbsoluteSeek 0
```

Comment:
Why is a seek needed here? Does `hSetFileSize` only truncate from the current position?

Reply:
That's copied form code I wrote when I introduced journaling. `hSetFileSize` does not change the file cursor, which means that the first write that follows will fill up the file with `\NUL` bytes up to the cursor, and then proceed to write normally.
Therefore, we need to manually seek to the start of the file.
In hindsight, that fact could've been documented better, but it seems we were a bit in a hurry back then when we fixed it, because it was rapidly filling the disk in production :see_no_evil:

Reply:
Ah, that explanation makes sense, but it would be nice to have that in a comment here indeed.

Reply:
Thanks for pointing that out. I have deduplicated this code now and added your explanation.